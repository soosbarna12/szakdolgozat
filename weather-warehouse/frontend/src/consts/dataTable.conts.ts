import { GridColDef } from "@mui/x-data-grid";

export const tempHistoricalDataTableColumns: GridColDef[] = [
  { field: "date", headerName: "Date", flex: 1 },
  { field: "cityName", headerName: "City Name", flex: 1 },
  { field: "countryCode", headerName: "Country", flex: 1 },
  { field: "maxTemp", headerName: "Max Temp (째C)", flex: 1 },
  { field: "minTemp", headerName: "Min Temp (째C)", flex: 1 },
];

export const allHistoricalDataTableColumns: GridColDef[] = [
  { field: "date", headerName: "Date", flex: 1 },
  { field: "cityName", headerName: "City Name", flex: 1 },
  { field: "countryCode", headerName: "Country", flex: 1 },
  { field: "maxTemp", headerName: "Max Temp (째C)", flex: 1 },
  { field: "minTemp", headerName: "Min Temp (째C)", flex: 1 },
  { field: "windSpeed", headerName: "Wind Speed (km/h)", flex: 1 },
  { field: "precipitation", headerName: "Precipitation (mm/s)", flex: 1 },
  { field: "pressure", headerName: "Pressure (hPa)", flex: 1 },
];