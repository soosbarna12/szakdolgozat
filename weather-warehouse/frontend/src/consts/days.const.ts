import { Day } from '../types/day.type';

export const DAYS: Array<Day> = [
	{
		id: 'MON',
		text: 'Monday'
	},
	{
		id: 'TUE',
		text: 'Tuesday'
	},
	{
		id: 'WED',
		text: 'Wednesday'
	},
	{
		id: 'THU',
		text: 'Thursday'
	},
	{
		id: 'FRI',
		text: 'Friday'
	},
	{
		id: 'SAT',
		text: 'Saturday'
	},
	{
		id: 'SUN',
		text: 'Sunday'
	}
];
