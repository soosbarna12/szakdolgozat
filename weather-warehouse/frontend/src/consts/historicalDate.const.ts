export const LOCAL_STORAGE_SELECTED_DATE_NAME = 'selectedDate';
