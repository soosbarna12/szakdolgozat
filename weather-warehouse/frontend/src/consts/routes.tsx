import React from 'react';
import { ForecastPage } from '../pages/ForecastPage/ForecastPage';
import { HistoricalPage } from '../pages/HistoricalPage/HistoricalPage';
import { TodayPage } from '../pages/TodayPage/TodayPage';
import { Pages } from '../types/page.type';

export const ROUTES = [
	{
		text: 'Historical',
		path: '/historical',
		id: Pages.Historical,
		element: <HistoricalPage />
	},
	{
		text: 'Today',
		path: '/today',
		id: Pages.Today,
		element: <TodayPage />
	},
	{
		text: 'Forecast',
		path: '/forecast',
		id: Pages.Forecast,
		element: <ForecastPage />
	}
];
