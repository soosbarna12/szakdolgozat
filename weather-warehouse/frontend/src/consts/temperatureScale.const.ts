
export const LOCAL_STORAGE_TEMPERATURE_SCALE = "temperatureScale";
