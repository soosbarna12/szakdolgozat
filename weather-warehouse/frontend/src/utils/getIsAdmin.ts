import { jwtDecode } from "jwt-decode";
import { JwtPayload } from "../types/jwtPayload";


export function getIsAdmin(): boolean {
    const token = localStorage.getItem('token');
    if (!token) return false;
    try {
      const decoded = jwtDecode<JwtPayload>(token);
      return decoded.role === 'admin';
    } catch {
      return false;
    }
  }