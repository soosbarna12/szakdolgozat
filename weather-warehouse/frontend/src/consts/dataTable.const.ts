import { GridColDef } from "@mui/x-data-grid";
import { TemperatureScale } from "../types/temperatureScale.type";
import { temperatureScaleChanger } from "../utils/temperatureScaleChanger";


export function getAllHistoricalDataTableColumns(tempScale: TemperatureScale): GridColDef[] {
  const temperatureScaleLabel = (tempScale === TemperatureScale.Celsius ? "°C" : "°F");

  return [
  { field: "date", headerName: "Date", flex: 1 },
  { field: "cityName", headerName: "City Name", flex: 1 },
  { field: "countryCode", headerName: "Country", flex: 1 },
  { field: "maxTemp", headerName: `Max Temp ${temperatureScaleLabel}`, flex: 1, valueFormatter: (value) => temperatureScaleChanger(TemperatureScale.Celsius, tempScale, value) },
  { field: "minTemp", headerName: `Min Temp ${temperatureScaleLabel}`, flex: 1, valueFormatter: (value) => temperatureScaleChanger(TemperatureScale.Celsius, tempScale, value) },
  { field: "windSpeed", headerName: "Wind Speed (km/h)", flex: 1 },
  { field: "precipitation", headerName: "Precipitation (mm/s)", flex: 1 },
  { field: "pressure", headerName: "Pressure (hPa)", flex: 1 },
]};