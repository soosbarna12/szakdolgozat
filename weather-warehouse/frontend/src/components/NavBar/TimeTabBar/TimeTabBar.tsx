import { Box, Container, Toolbar, useTheme } from '@mui/material';
import React from 'react';
import { Link, useLocation } from 'react-router-dom';
import { ROUTES } from '../../../consts/routes';
import { StyledTimeTabButton } from '../../../stlyes/button.style';

export function TimeTabBar() {
  const location = useLocation();
  const theme = useTheme();

  // Check if the logged-in user is an admin
  const isAdmin = localStorage.getItem('loggedIn') === 'true' && localStorage.getItem('role') === 'admin';

  function renderMenu() {
    return ROUTES.filter((route) => {
      // Exclude the "Admin" route if the user is not an admin
      if (route.id === 'Admin' && !isAdmin) {
        return false;
      }
      return true;
    }).map((route) => {
      const isActive = location.pathname === route.path;
      return (
        <StyledTimeTabButton
          component={Link}
          to={route.path}
          key={route.id}
          sx={{
            boxShadow: 4,
            backgroundColor: isActive
              ? theme.palette.primary.dark
              : theme.palette.background.paper,
            color: isActive ? '#fff' : 'inherit',
          }}
        >
          {route.text}
        </StyledTimeTabButton>
      );
    });
  }

  return (
    <Container sx={{ display: 'flex', justifyContent: 'center' }}>
      <Toolbar disableGutters>
        <Box>{renderMenu()}</Box>
      </Toolbar>
    </Container>
  );
}
