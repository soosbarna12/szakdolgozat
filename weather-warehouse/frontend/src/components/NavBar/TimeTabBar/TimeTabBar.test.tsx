import '@testing-library/jest-dom';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter, useLocation } from 'react-router-dom';
import { ROUTES } from '../../../consts/routes';
import { getIsAdmin } from '../../../utils/getIsAdmin';
import { renderWithQueryClient } from '../../../utils/test/renderWithQueryClient';
import { TimeTabBar } from './TimeTabBar';


jest.mock('../../../utils/getIsAdmin', () => ({
    getIsAdmin: jest.fn(),
}));

jest.mock('react-leaflet', () => ({
    MapContainer: ({ children }: { children: React.ReactNode }) => (
        <div data-testid="dataMapContainer">{children}</div>
    )
}));

jest.mock('export-to-csv', () => ({
    download: jest.fn(),
    generateCsv: jest.fn(),
    mkConfig: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: jest.fn(),
}));

const renderTimeTabBar = () =>
    renderWithQueryClient(
        <MemoryRouter>
            <TimeTabBar />
        </MemoryRouter>
    );


describe('NavBar/TimeTabBar', () => {
    const mockUseLocation = useLocation as jest.Mock;
    const mockGetIsAdmin = getIsAdmin as jest.Mock;

    beforeEach(() => {
        jest.clearAllMocks();
        mockUseLocation.mockReturnValue({ pathname: '/today' });
        mockGetIsAdmin.mockReturnValue(false);
    });

    it('matches TimeTabBar component snapshot', () => {
        const component = renderTimeTabBar();
        expect(component).toMatchSnapshot();
    });

    it('renders all tabs and highlights the selected tab', () => {
        renderTimeTabBar();

        ROUTES.forEach((route) => {
            if (route.id === 'Admin') {
                expect(screen.queryByTestId(`tab${route.id}`)).not.toBeInTheDocument();
            } else {
                const tab = screen.getByTestId(`tab${route.id}`);
                expect(tab).toBeInTheDocument();
                if (route.path === '/today') {
                    expect(tab).toHaveStyle('background-color: rgb(21, 101, 192)'); // Updated to match the actual value
                } else {
                    expect(tab).toHaveStyle('background-color: rgb(255, 255, 255)'); // Assuming background.paper is white
                }
            }
        });
    });

    it('renders the Admin tab if the user is an admin', () => {
        mockGetIsAdmin.mockReturnValue(true);
        renderTimeTabBar();

        const adminTab = screen.getByTestId('tabAdmin');
        expect(adminTab).toBeInTheDocument();
    });

    it('navigates to the correct route when a tab is clicked', async () => {
        renderTimeTabBar();

        const historicalTab = screen.getByTestId('tabHistorical');
        fireEvent.click(historicalTab);

        await waitFor(() => {
            expect(mockUseLocation).toHaveBeenCalledWith();
        });
    });
});