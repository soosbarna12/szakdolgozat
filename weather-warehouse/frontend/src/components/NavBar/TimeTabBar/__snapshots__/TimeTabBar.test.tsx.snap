// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar/TimeTabBar matches TimeTabBar component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-8y1sfv-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1u9o2lj-MuiToolbar-root"
        >
          <div
            class="MuiBox-root css-0"
          >
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-ho9ssy-MuiButtonBase-root-MuiButton-root"
              data-discover="true"
              data-testid="tabHistorical"
              href="/historical"
              tabindex="0"
            >
              Historical
            </a>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1atnkzf-MuiButtonBase-root-MuiButton-root"
              data-discover="true"
              data-testid="tabToday"
              href="/today"
              tabindex="0"
            >
              Today
            </a>
            <a
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-ho9ssy-MuiButtonBase-root-MuiButton-root"
              data-discover="true"
              data-testid="tabForecast"
              href="/forecast"
              tabindex="0"
            >
              Forecast
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-8y1sfv-MuiContainer-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular css-1u9o2lj-MuiToolbar-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-ho9ssy-MuiButtonBase-root-MuiButton-root"
            data-discover="true"
            data-testid="tabHistorical"
            href="/historical"
            tabindex="0"
          >
            Historical
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1atnkzf-MuiButtonBase-root-MuiButton-root"
            data-discover="true"
            data-testid="tabToday"
            href="/today"
            tabindex="0"
          >
            Today
          </a>
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-ho9ssy-MuiButtonBase-root-MuiButton-root"
            data-discover="true"
            data-testid="tabForecast"
            href="/forecast"
            tabindex="0"
          >
            Forecast
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
